### PR TITLE
use version check for disabling global CLI

### PR DESCRIPTION
**Pull request by EvanBacon**
# Checklist

- [ ] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary. You can comment this pull request with `/changelog-entry [breaking-change|new-feature|bug-fix|chore] [message]` and CHANGELOG.md will be updated automatically.

# Why

- revert 51db45a8

# Test Plan

- Added tests

### CODE REVIEW

**Comment by EvanBacon** [comment]
/changelog-entry new-feature use local CLI for `expo export` in SDK +46

**Review by brentvatne — APPROVED** [review]

File: packages/eas-cli/src/utils/expoCli.ts [hunk 1]
```diff
@@ -1,8 +1,60 @@
...
+  // Run the environment variable check first as it's the cheapest.
+  const userDefinedVersionedCliEnabled = boolish('EXPO_USE_LOCAL_CLI', true);
+  if (!userDefinedVersionedCliEnabled) {
+    return false;
+  }
+
+  try {
+    // NOTE(EvanBacon): The CLI package could be available through a monorepo
+    // we need to ensure the project is specifically using a known supported Expo SDK version.
+    const { sdkVersion } = getConfig(projectDir).exp;
```
Comment (dsokal) [inline packages/eas-cli/src/utils/expoCli.ts:39]: Don't call `getConfig` explicitly. This should be passed from the command's context. Talk with @wschurman if you need details :)

File: packages/eas-cli/src/utils/expodash/memoize.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  const cache: { [key: string]: any } = {};
+  return ((...args: any[]) => {
+    const key = JSON.stringify(args);
+    if (cache[key]) {
+      return cache[key];
+    }
+    const result = fn(...args);
+    cache[key] = result;
+    return result;
+  }) as any;
```
Comment (dsokal) [inline packages/eas-cli/src/utils/expodash/memoize.ts:12]: 🤔 
```suggestion
  }) as T;
```

**Review by dsokal — CHANGES_REQUESTED** [review]
Requesting changes because of the `getConfig` comment.

File: packages/eas-cli/src/utils/expoCli.ts [hunk 3]
```diff
@@ -1,8 +1,60 @@
...
+function gteSdkVersion(fromSdkVersion: string, sdkVersion: string): boolean {
+  if (fromSdkVersion === 'UNVERSIONED') {
+    return true;
+  }
+
+  try {
+    return semver.gte(fromSdkVersion, sdkVersion);
+  } catch {
+    return true;
+  }
```
[the remark this answers is not in the document]
Comment (EvanBacon) [reply packages/eas-cli/src/utils/expoCli.ts:21]: yes

File: CHANGELOG.md [hunk 4]
```diff
@@ -9,6 +9,7 @@ This is the log of notable changes to EAS CLI and related packages.
 ### 🎉 New features
 
 - Add custom metadata validation command with more complex rules. ([#1416](https://github.com/expo/eas-cli/pull/1416) by [@byCedric](https://github.com/byCedric))
+- Use local CLI for `expo export` in SDK +46. ([#1474](https://github.com/expo/eas-cli/pull/1474) by [@EvanBacon](https://github.com/EvanBacon))
```
Comment (dsokal) [inline CHANGELOG.md:12]: Rebase and move to the unreleased section.

File: packages/eas-cli/src/utils/expoCli.ts [hunk 5]
```diff
@@ -1,8 +1,60 @@
...
+  // Run the environment variable check first as it's the cheapest.
+  const userDefinedVersionedCliEnabled = boolish('EXPO_USE_LOCAL_CLI', true);
+  if (!userDefinedVersionedCliEnabled) {
+    return false;
+  }
+
+  try {
+    // NOTE(EvanBacon): The CLI package could be available through a monorepo
+    // we need to ensure the project is specifically using a known supported Expo SDK version.
+    const { sdkVersion } = getExpoConfig(projectDir);
```
Comment (dsokal) [inline packages/eas-cli/src/utils/expoCli.ts:39]: My comment is still not addressed.
The ExpoConfig object should be passed from the top-level code (publish/index.ts):

```diff
diff --git a/packages/eas-cli/src/commands/update/index.ts b/packages/eas-cli/src/commands/update/index.ts
index abdde689..9ae47247 100644
--- a/packages/eas-cli/src/commands/update/index.ts
+++ b/packages/eas-cli/src/commands/update/index.ts
@@ -391,7 +391,7 @@ export default class UpdatePublish extends EasCommand {
       if (!skipBundler) {
         const bundleSpinner = ora().start('Building bundle...');
         try {
-          await buildBundlesAsync({ projectDir, inputDir });
+          await buildBundlesAsync({ exp, projectDir, inputDir });
           bundleSpinner.succeed('Built bundle!');
         } catch (e) {
           bundleSpinner.fail('Failed to build bundle!');
diff --git a/packages/eas-cli/src/project/publish.ts b/packages/eas-cli/src/project/publish.ts
index 7726e3b0..a12273b2 100644
--- a/packages/eas-cli/src/project/publish.ts
+++ b/packages/eas-cli/src/project/publish.ts
@@ -146,9 +146,11 @@ export async function buildUnsortedUpdateInfoGroupAsync(
 }
 
 export async function buildBundlesAsync({
+  exp,
   projectDir,
   inputDir,
 }: {
+  exp: ExpoConfig;
   projectDir: string;
   inputDir: string;
 }): Promise<void> {
@@ -157,7 +159,7 @@ export async function buildBundlesAsync({
     throw new Error('Could not locate package.json');
   }
 
-  if (shouldUseVersionedExpoCLI(projectDir)) {
+  if (shouldUseVersionedExpoCLI(projectDir, exp)) {
     await expoCommandAsync(projectDir, ['export', '--output-dir', inputDir, '--dump-sourcemap']);
   } else {
     // Legacy global Expo CLI
diff --git a/packages/eas-cli/src/utils/expoCli.ts b/packages/eas-cli/src/utils/expoCli.ts
index 4a2342eb..dc7a3912 100644
--- a/packages/eas-cli/src/utils/expoCli.ts
+++ b/packages/eas-cli/src/utils/expoCli.ts
@@ -1,3 +1,4 @@
+import { ExpoConfig } from '@expo/config-types';
 import spawnAsync from '@expo/spawn-async';
 import chalk from 'chalk';
 import { boolish } from 'getenv';
@@ -5,7 +6,6 @@ import resolveFrom from 'resolve-from';
 import semver from 'semver';
 
 import Log from '../log';
-import { getExpoConfig } from '../project/expoConfig';
 import { memoize } from './expodash/memoize';
 
 // Aggressively returns `true` (UNVERSIONED, invalid SDK version format) to push users towards the versioned CLI.
@@ -24,7 +24,7 @@ function gteSdkVersion(fromSdkVersion: string, sdkVersion: string): boolean {
 /**
  * @returns `true` if the project is SDK +46, has `@expo/cli`, and `EXPO_USE_LOCAL_CLI` is not set to a _false_ value.
  */
-export function shouldUseVersionedExpoCLIExpensive(projectDir: string): boolean {
+export function shouldUseVersionedExpoCLIExpensive(projectDir: string, exp: ExpoConfig): boolean {
   // Users can disable local CLI settings EXPO_USE_LOCAL_CLI=false
   // https://github.com/expo/expo/blob/69eddda7bb1dbfab44258f468cf7f22984c1e44e/packages/expo/bin/cli.js#L10
   // Run the environment variable check first as it's the cheapest.
@@ -36,7 +36,7 @@ export function shouldUseVersionedExpoCLIExpensive(projectDir: string): boolean
   try {
     // NOTE(EvanBacon): The CLI package could be available through a monorepo
     // we need to ensure the project is specifically using a known supported Expo SDK version.
-    const { sdkVersion } = getExpoConfig(projectDir);
+    const { sdkVersion } = exp;
 
     if (
       // If the version isn't defined then skip the check.
```

**Review by dsokal — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 4 above)
Comment (dsokal) [reply CHANGELOG.md:12]: This is not addressed. You haven't rebased your branch. I'll fix it in main.

**Pull request merged**